File: packages/oauth/oauth-client-expo/android/src/main/java/expo/modules/atprotooauthclient/ExpoAtprotoOAuthClientModule.kt
package expo.modules.atprotooauthclient

import expo.modules.atprotooauthclient.Crypto
import expo.modules.atprotooauthclient.Jose
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoAtprotoOAuthClientModule : Module() {
  override fun definition() =
    ModuleDefinition {
      Name("ExpoAtprotoOAuthClient")

      AsyncFunction("digest") { data: ByteArray, algo: String ->
        if (algo != "sha256") {
          throw IllegalArgumentException("Unsupported algorithm: $algo")
        }
        return@AsyncFunction Crypto().digest(data)
      }

      AsyncFunction("getRandomValues") { byteLength: Int ->
        return@AsyncFunction Crypto().getRandomValues(byteLength)
      }

      AsyncFunction("generatePrivateJwk") { algo: String ->
        if (algo != "ES256") {
          throw IllegalArgumentException("Unsupported algorithm: $algo")
        }
        return@AsyncFunction Crypto().generateJwk()
      }

      AsyncFunction("createJwt") { header: String, payload: String, encodedJwk: EncodedJWK ->
        val jwk = Crypto().decodeJwk(encodedJwk)
        return@AsyncFunction Jose().createJwt(header, payload, jwk)
      }

      AsyncFunction("verifyJwt") { token: String, encodedJwk: EncodedJWK, options: VerifyOptions ->
        val jwk = Crypto().decodeJwk(encodedJwk)
        return@AsyncFunction Jose().verifyJwt(token, jwk, options)
      }
    }
}

File: packages/oauth/oauth-client-expo/ios/ExpoAtprotoOAuthClientModule.swift
import ExpoModulesCore

enum ExpoAtprotoOAuthClientError: Error {
  case unsupportedAlgorithm(String)
  case invalidJwk
  case invalidHeader(String)
  case invalidPayload(String)
  case nullSigner
}

public class ExpoAtprotoOAuthClientModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoAtprotoOAuthClient")

    AsyncFunction("digest") { (data: Data, algo: String) throws -> Data in
      if algo != "sha256" {
        throw ExpoAtprotoOAuthClientError.unsupportedAlgorithm(algo)
      }
      return CryptoUtil.digest(data: data)
    }

    AsyncFunction("getRandomValues") { (byteLength: Int) -> Data in
      return CryptoUtil.getRandomValues(byteLength: byteLength)
    }

    AsyncFunction("generatePrivateJwk") { (algo: String) throws -> EncodedJWK in
      if algo != "ES256" {
        throw ExpoAtprotoOAuthClientError.unsupportedAlgorithm(algo)
      }
      return CryptoUtil.generateJwk()
    }

    AsyncFunction("createJwt") { (header: String, payload: String, jwk: EncodedJWK) throws -> String in
      let jwk = try CryptoUtil.decodeJwk(x: jwk.x, y: jwk.y, d: jwk.d)
      let jwt = try JoseUtil.createJwt(header: header, payload: payload, jwk: jwk)
      return jwt
    }

    AsyncFunction("verifyJwt") { (token: String, jwk: EncodedJWK, options: VerifyOptions) throws -> VerifyResult in
      let jwk = try CryptoUtil.decodeJwk(x: jwk.x, y: jwk.y, d: jwk.d)
      let res = try JoseUtil.verifyJwt(token: token, jwk: jwk, options: options)
      return res
    }
  }
}
